### PR TITLE
Enabled CI for Windows

### PR DESCRIPTION
**Pull request by UebelAndre**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?
This change includes all the necessary fixes and updates to support enabling CI for windows platforms.

- [ ] Bugfix
- [x] Feature (please, look at the "Scope of the project" section in the README.md file)
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: https://github.com/bazelbuild/rules_python/issues/2
I don't know if this closes https://github.com/bazelbuild/rules_python/issues/2 but seems related

## What is the new behavior?
Updated rules and tests to support windows platforms.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Other information
closes https://github.com/bazelbuild/rules_python/pull/572

### CODE REVIEW

**Comment by UebelAndre** [comment]
@alexeagle you may want to take a look at the changes to the integration test runner. I could be mistaken but I think you have the most context on that bit of code 😅

File: docs/BUILD [hunk 1]
```diff
@@ -83,6 +83,13 @@ stardoc(
     name = "core-docs",
     out = "python.md_",
     input = "//python:defs.bzl",
+    # TODO: Stardoc does not guarantee consistent outputs accross platforms (Unix/Windows).
+    # As a result we do not build or test docs on Windows.
+    target_compatible_with = select({
```
Comment (alexeagle) [inline docs/BUILD:88]: nit: I'd pull the select out to a constant named _NOT_WINDOWS

File: python/pip_install/extract_wheels/lib/BUILD [hunk 2]
```diff
@@ -70,18 +70,22 @@ py_test(
...
-    srcs = [
-        "whl_filegroup_test.py",
-    ],
-    data = ["//examples/wheel:minimal_with_py_package"],
-    tags = ["unit"],
-    deps = [
-        ":lib",
-    ],
-)
+# The `whl_filegroup_test` changes directories at runtime which impacts the
```
Comment (alexeagle) [inline python/pip_install/extract_wheels/lib/BUILD:73]: could we just fix the test to restore the original conditions when it finishes, so it's safe to run them in the same interpreter?

File: python/pip_install/extract_wheels/lib/BUILD [hunk 3]
```diff
@@ -70,18 +70,22 @@ py_test(
...
+        name = "whl_filegroup_{}_test".format(incremental),
+        size = "small",
+        srcs = ["whl_filegroup_test.py"],
+        data = ["//examples/wheel:minimal_with_py_package"],
+        env = {"INCREMENTAL": incremental},
+        main = "whl_filegroup_test.py",
+        tags = ["unit"],
+        deps = [":lib"],
+    )
+    for incremental in ("incremental", "non_incremental")
```
Comment (alexeagle) [inline python/pip_install/extract_wheels/lib/BUILD:87]: or, this seems like a manual sharding scheme, we could just set a shard_count to get them isolated

File: python/pip_install/extract_wheels/lib/whl_filegroup_test.py [hunk 4]
```diff
@@ -33,19 +30,29 @@ def _run(
...
+        generated_bazel_dir = (
+            generated_bazel_dir[2:] if not incremental else self.wheel_dir
+        )
         self.assertIn(self.wheel_name, os.listdir(generated_bazel_dir))
         with open("{}/BUILD.bazel".format(generated_bazel_dir)) as build_file:
             build_file_content = build_file.read()
-            self.assertIn('filegroup', build_file_content)
+            self.assertIn("filegroup", build_file_content)
 
+    @unittest.skipIf(
```
Comment (alexeagle) [inline python/pip_install/extract_wheels/lib/whl_filegroup_test.py:45]: clever, but also kinda non-obvious

File: python/pip_install/extract_wheels/lib/BUILD (hunk 2 above)
Comment (UebelAndre) [reply python/pip_install/extract_wheels/lib/BUILD:73]: The test is lazily written IMO. `pwd` should not be a requirement for this functionality and patterns where you need to `chdir` are almost never maintainable. I think the correct fix is to refactor these all together but this was the quickest path forward for me. I could remove the sharding and the tests would work but there'd be a potential flake. I'd rather do that than a larger refactor.

Comment (UebelAndre) [reply python/pip_install/extract_wheels/lib/BUILD]: Ah, this ended up being (potentially) small 😄

File: python/pip_install/extract_wheels/lib/BUILD (hunk 3 above)
Comment (UebelAndre) [reply python/pip_install/extract_wheels/lib/BUILD]: refactored this to not be necessary.

**Comment by UebelAndre** [comment]
@alexeagle all comments addressed and ready for another review 😄

File: tools/bazel_integration_test/test_runner.py [hunk 5]
```diff
@@ -25,9 +26,12 @@ def modify_WORKSPACE(wksp, distro_path):
     with open(wksp, "r") as wksp_file:
         content = wksp_file.read()
     # Replace the url for rules_python with our locally built one
+    distro_runfile_path = r.Rlocation(distro_path)
+    if platform.system() == "Windows":
+        distro_runfile_path = "/" + distro_runfile_path.replace("\\", "/")
     content = re.sub(
         r'url = "https://github.com/bazelbuild/rules_python/[^"]+"',
-        'url = "file://%s"' % r.Rlocation(distro_path),
+        'url = "file://%s"' % distro_runfile_path,
```
Comment (alexeagle) [inline tools/bazel_integration_test/test_runner.py:34]: I'm about to throw away the distro file, just FYI.

File: python/pip_install/extract_wheels/lib/whl_filegroup_test.py [hunk 6]
```diff
@@ -31,12 +29,14 @@ def _run(
             enable_implicit_namespace_pkgs=False,
             incremental=incremental,
             repo_prefix=repo_prefix,
+            incremental_dir=Path(self.wheel_dir),
```
Comment (alexeagle) [inline python/pip_install/extract_wheels/lib/whl_filegroup_test.py:32]: thanks, much nicer way to ensure tests are isolated from each other

**Comment by alexeagle** [comment]
Does this fix https://github.com/bazelbuild/rules_python/issues/2 ? if not, do you know what's left?

**Comment by UebelAndre** [comment]
> Does this fix #2 ? if not, do you know what's left?

I'm not entirely sure. I just recalled seeing a "Add windows support" style issue and thought it'd be relevant to this.

**Comment by UebelAndre** [comment]
This PR would also close out https://github.com/bazelbuild/rules_python/pull/344

**Review by alexeagle — APPROVED** [review]
awesome, thanks!

**Pull request merged**